File: tools/retro-muri.py
#!/usr/bin/env python3

# retro-muri is an assembler for Nga, the virtual machine at the heart of
# Retro. It is used to build the image file containing the actual Retro
# language.
#
# This will extract the code blocks in the specified file and generate an
# image file named `ngaImage`.
#
# Copyright (c) 2020, Charles Childers
#
# Usage:
#
#    retro.muri.py filename

import sys, struct

# labels stores the label names as a dictionary, with the key being
# the label name and the value being the location in memory.
#
# image stores the assembled opcodes and data.

labels = dict()
IMAGE_SIZE = 128 * 1024
image = [0] * IMAGE_SIZE
here = 0
last_dictionary_entry = 0
dictionary_entries = 0

INSTRUCTIONS = [
    "..", "li", "du", "dr", "sw", "pu", "po", "ju", "ca", "cc",
    "re", "eq", "ne", "lt", "gt", "fe", "st", "ad", "su", "mu",
    "di", "an", "or", "xo", "sh", "zr", "ha", "ie", "iq", "ii",
]
IP_MODIFYING_INSTRUCTIONS = {"ju", "ca", "cc", "re", "zr"}


def strip_comments(line):
    """Strip Pali-style trailing comments, except from string data."""
    if line.startswith("s"):
        return line

    hash_pos = line.find("#")
    while hash_pos != -1:
        if hash_pos == 0 or line[hash_pos - 1] in " \t":
            return line[:hash_pos].rstrip(" \t")
        hash_pos = line.find("#", hash_pos + 1)
    return line.rstrip(" \t")


def validate_instruction_bundle(line, line_number):
    """Validate an `i` directive against the documented Nga instruction rules."""
    if not line.startswith("i "):
        return

    bundle = line[2:]
    if len(bundle) != 8:
        raise ValueError(
            f"Error on line {line_number}: instruction bundle must contain "
            "exactly four two-character instructions"
        )

    instructions = [bundle[i : i + 2] for i in range(0, 8, 2)]
    for instruction in instructions:
        if instruction not in INSTRUCTIONS:
            raise ValueError(
                f"Error on line {line_number}: invalid Nga instruction "
                f"'{instruction}' in bundle '{bundle}'"
            )

    for index, instruction in enumerate(instructions[:-1]):
        if instruction in IP_MODIFYING_INSTRUCTIONS:
            if any(following != ".." for following in instructions[index + 1 :]):
                raise ValueError(
                    f"Error on line {line_number}: Nga instruction "
                    f"'{instruction}' must be followed only by NOPs (..) "
                    "in its bundle"
                )

# assemble() takes a string representation of an opcode bundle,
# finds the individual opcodes, packs them into a cell-sized value,
# and returns this value.
#
# Each instruction bundle has four two character instruction names,
# with `..` used to represent a non-operation instruction.


def assemble(inst):
    a = INSTRUCTIONS.index(inst[0:2])
    b = INSTRUCTIONS.index(inst[2:4])
    c = INSTRUCTIONS.index(inst[4:6])
    d = INSTRUCTIONS.index(inst[6:8])
    o = int.from_bytes([a, b, c, d], byteorder="little", signed=False)
    return o


# Each pass extracts Muri code blocks and applies the supplied handler.


def process_source(handler):
    f = sys.argv[1]
    in_block = False
    with open(f, "r") as source:
        for line_number, raw_line in enumerate(source.readlines(), start=1):
            if raw_line.rstrip() == "~~~":
                in_block = not in_block
            elif in_block:
                line = strip_comments(raw_line.rstrip("\n"))
                validate_instruction_bundle(line, line_number)
                if not line or line[0] == "c":
                    continue
                handler(line)


def parse_dict_line(line):
    fields = [field for field in line[2:].replace("\t", " ").split(" ") if field]
    if len(fields) != 3:
        raise ValueError("Dictionary entries require a name, label, and class handler")
    return fields


def dict_entry_size(name):
    return 9 + len(name.encode("utf-8")) + 1


def lookup(name):
    return labels.get(name, -1)


# The first pass records label locations and lays out every directive.


def pass1():
    global here
    here = 0

    def assemble_line(line):
        global here
        directive = line[0]
        if directive in "ir-d":
            here += 1
        elif directive == "s":
            here += len(line[2:].encode("utf-8")) + 1
        elif directive == "o":
            here = int(line[2:])
        elif directive == "*":
            here += int(line[2:])
        elif directive == "D":
            name, _, _ = parse_dict_line(line)
            here += dict_entry_size(name)
        elif directive == ":":
            name = line[2:]
            if lookup(name) != -1:
                raise ValueError(f"Fatal error: {name} already defined")
            labels[name] = here

    process_source(assemble_line)


# The second through fifth passes respectively emit opcodes, numeric data,
# strings and dictionary names, and references and dictionary headers.


def pass2():
    global here
    here = 0

    def assemble_line(line):
        global here
        directive = line[0]
        if directive == "i":
            image[here] = assemble(line[2:])
            here += 1
        elif directive == "o":
            here = int(line[2:])
        elif directive == "*":
            here += int(line[2:])
        elif directive == "D":
            name, _, _ = parse_dict_line(line)
            here += dict_entry_size(name)
        elif directive in "-rds":
            here += 1 if directive != "s" else len(line[2:].encode("utf-8")) + 1

    process_source(assemble_line)


def pass3():
    global here
    here = 0

    def assemble_line(line):
        global here
        directive = line[0]
        if directive in "ir-":
            here += 1
        elif directive == "o":
            here = int(line[2:])
        elif directive == "*":
            here += int(line[2:])
        elif directive == "d":
            image[here] = int(line[2:])
            here += 1
        elif directive == "D":
            name, _, _ = parse_dict_line(line)
            here += dict_entry_size(name)
        elif directive == "s":
            here += len(line[2:].encode("utf-8")) + 1

    process_source(assemble_line)


def pass4():
    global here
    here = 0

    def assemble_line(line):
        global here
        directive = line[0]
        if directive in "ir-d":
            here += 1
        elif directive == "o":
            here = int(line[2:])
        elif directive == "*":
            here += int(line[2:])
        elif directive == "s":
            for byte in line[2:].encode("utf-8"):
                image[here] = byte
                here += 1
            image[here] = 0
            here += 1
        elif directive == "D":
            start = here
            name, _, _ = parse_dict_line(line)
            name_bytes = name.encode("utf-8")
            for offset, byte in enumerate(name_bytes):
                image[start + 9 + offset] = byte
            image[start + 9 + len(name_bytes)] = 0
            here += dict_entry_size(name)

    process_source(assemble_line)


def pass5():
    global here, last_dictionary_entry, dictionary_entries
    here = 0
    last_dictionary_entry = 0
    dictionary_entries = 0

    def assemble_line(line):
        global here, last_dictionary_entry, dictionary_entries
        directive = line[0]
        if directive == "i":
            here += 1
        elif directive == "o":
            here = int(line[2:])
        elif directive == "*":
            here += int(line[2:])
        elif directive in "r-":
            name = line[2:]
            image[here] = lookup(name)
            if image[here] == -1:
                print(f"Lookup failed: '{name}'")
            here += 1
        elif directive == "d":
            here += 1
        elif directive == "s":
            here += len(line[2:].encode("utf-8")) + 1
        elif directive == "D":
            start = here
            name, label, class_name = parse_dict_line(line)
            size = dict_entry_size(name)
            xt = lookup(label)
            class_handler = lookup(class_name)
            if xt == -1:
                print(f"Lookup failed: '{label}'")
            if class_handler == -1:
                print(f"Lookup failed: '{class_name}'")
            image[start : start + 9] = [
                0 if dictionary_entries == 0 else last_dictionary_entry,
                xt, class_handler, 0, 0, 0, 0, 0, 0,
            ]
            last_dictionary_entry = start
            dictionary_entries += 1
            here += size

    process_source(assemble_line)


# save() handles writing the image to a file


def save(filename):
    with open(filename, "wb") as file:
        for cell in image[:here]:
            file.write(struct.pack("=i", cell))


if __name__ == "__main__":
    try:
        pass1()
        pass2()
        pass3()
        pass4()
        pass5()
        save("ngaImage")
    except ValueError as error:
        print(error)
        sys.exit(1)

File: tools/retro-muri.js
#!/usr/bin/env node

/* RETRO ------------------------------------------------------
   A Node.js implementation of the standard Nga assembler.
   It reads Muri code blocks and writes an image named ngaImage.
   ---------------------------------------------------------- */

"use strict";

const fs = require("fs");

const IMAGE_SIZE = 128 * 1024;
const INSTRUCTIONS = [
  "..", "li", "du", "dr", "sw", "pu", "po", "ju", "ca", "cc",
  "re", "eq", "ne", "lt", "gt", "fe", "st", "ad", "su", "mu",
  "di", "an", "or", "xo", "sh", "zr", "ha", "ie", "iq", "ii",
];
const IP_MODIFYING_INSTRUCTIONS = new Set(["ju", "ca", "cc", "re", "zr"]);

const labels = new Map();
const image = new Int32Array(IMAGE_SIZE);
let here = 0;
let lastDictionaryEntry = 0;
let dictionaryEntries = 0;

function stripComments(line) {
  if (line[0] === "s") {
    return line;
  }

  let hashPosition = line.indexOf("#");
  while (hashPosition !== -1) {
    if (hashPosition === 0 || line[hashPosition - 1] === " " || line[hashPosition - 1] === "\t") {
      return line.slice(0, hashPosition).replace(/[ \t]+$/, "");
    }
    hashPosition = line.indexOf("#", hashPosition + 1);
  }
  return line.replace(/[ \t]+$/, "");
}

function validateInstructionBundle(line, lineNumber) {
  if (!line.startsWith("i ")) {
    return;
  }

  const bundle = line.slice(2);
  if (bundle.length !== 8) {
    throw new Error(`Error on line ${lineNumber}: instruction bundle must contain exactly four two-character instructions`);
  }

  const instructions = [];
  for (let index = 0; index < 8; index += 2) {
    const instruction = bundle.slice(index, index + 2);
    if (!INSTRUCTIONS.includes(instruction)) {
      throw new Error(`Error on line ${lineNumber}: invalid Nga instruction '${instruction}' in bundle '${bundle}'`);
    }
    instructions.push(instruction);
  }

  for (let index = 0; index < 3; index++) {
    if (IP_MODIFYING_INSTRUCTIONS.has(instructions[index])) {
      for (let following = index + 1; following < 4; following++) {
        if (instructions[following] !== "..") {
          throw new Error(`Error on line ${lineNumber}: Nga instruction '${instructions[index]}' must be followed only by NOPs (..) in its bundle`);
        }
      }
    }
  }
}

function processSource(filename, handler) {
  const lines = fs.readFileSync(filename, "utf8").split("\n");
  let inBlock = false;

  lines.forEach((rawLine, index) => {
    const lineNumber = index + 1;
    if (rawLine === "~~~") {
      inBlock = !inBlock;
    } else if (inBlock) {
      const line = stripComments(rawLine);
      validateInstructionBundle(line, lineNumber);
      if (line.length > 0 && line[0] !== "c") {
        handler(line);
      }
    }
  });
}

function parseAtoi(value) {
  const match = value.match(/^[ \t]*([+-]?\d+)/);
  return match === null ? 0 : Number.parseInt(match[1], 10);
}

function parseDictionaryLine(line) {
  const fields = line.slice(2).split(/[ \t]+/).filter((field) => field.length > 0);
  return [fields[0] || "", fields[1] || "", fields[2] || ""];
}

function dictionaryEntrySize(name) {
  return 9 + Buffer.byteLength(name, "utf8") + 1;
}

function lookup(name) {
  return labels.has(name) ? labels.get(name) : -1;
}

function assemble(bundle) {
  const a = INSTRUCTIONS.indexOf(bundle.slice(0, 2));
  const b = INSTRUCTIONS.indexOf(bundle.slice(2, 4));
  const c = INSTRUCTIONS.indexOf(bundle.slice(4, 6));
  const d = INSTRUCTIONS.indexOf(bundle.slice(6, 8));
  return (a | (b << 8) | (c << 16) | (d << 24)) >>> 0;
}

function pass1(filename) {
  here = 0;
  processSource(filename, (line) => {
    const directive = line[0];
    if ("ir-d".includes(directive)) {
      here++;
    } else if (directive === "s") {
      here += Buffer.byteLength(line.slice(2), "utf8") + 1;
    } else if (directive === "o") {
      here = parseAtoi(line.slice(2));
    } else if (directive === "*") {
      here += parseAtoi(line.slice(2));
    } else if (directive === "D") {
      const [name] = parseDictionaryLine(line);
      here += dictionaryEntrySize(name);
    } else if (directive === ":") {
      const name = line.slice(2);
      if (labels.has(name)) {
        throw new Error(`Fatal error: ${name} already defined`);
      }
      labels.set(name, here);
    }
  });
}

function pass2(filename) {
  here = 0;
  processSource(filename, (line) => {
    const directive = line[0];
    if (directive === "i") {
      image[here++] = assemble(line.slice(2));
    } else if (directive === "o") {
      here = parseAtoi(line.slice(2));
    } else if (directive === "*") {
      here += parseAtoi(line.slice(2));
    } else if (directive === "D") {
      const [name] = parseDictionaryLine(line);
      here += dictionaryEntrySize(name);
    } else if ("-rd".includes(directive)) {
      here++;
    } else if (directive === "s") {
      here += Buffer.byteLength(line.slice(2), "utf8") + 1;
    }
  });
}

function pass3(filename) {
  here = 0;
  processSource(filename, (line) => {
    const directive = line[0];
    if ("ir-".includes(directive)) {
      here++;
    } else if (directive === "o") {
      here = parseAtoi(line.slice(2));
    } else if (directive === "*") {
      here += parseAtoi(line.slice(2));
    } else if (directive === "d") {
      image[here++] = parseAtoi(line.slice(2));
    } else if (directive === "D") {
      const [name] = parseDictionaryLine(line);
      here += dictionaryEntrySize(name);
    } else if (directive === "s") {
      here += Buffer.byteLength(line.slice(2), "utf8") + 1;
    }
  });
}

function pass4(filename) {
  here = 0;
  processSource(filename, (line) => {
    const directive = line[0];
    if ("ir-d".includes(directive)) {
      here++;
    } else if (directive === "o") {
      here = parseAtoi(line.slice(2));
    } else if (directive === "*") {
      here += parseAtoi(line.slice(2));
    } else if (directive === "s") {
      for (const byte of Buffer.from(line.slice(2), "utf8")) {
        image[here++] = byte;
      }
      image[here++] = 0;
    } else if (directive === "D") {
      const start = here;
      const [name] = parseDictionaryLine(line);
      const nameBytes = Buffer.from(name, "utf8");
      nameBytes.forEach((byte, index) => { image[start + 9 + index] = byte; });
      image[start + 9 + nameBytes.length] = 0;
      here += dictionaryEntrySize(name);
    }
  });
}

function pass5(filename) {
  here = 0;
  lastDictionaryEntry = 0;
  dictionaryEntries = 0;
  processSource(filename, (line) => {
    const directive = line[0];
    if (directive === "i") {
      here++;
    } else if (directive === "o") {
      here = parseAtoi(line.slice(2));
    } else if (directive === "*") {
      here += parseAtoi(line.slice(2));
    } else if ("r-".includes(directive)) {
      const name = line.slice(2);
      image[here] = lookup(name);
      if (image[here] === -1) {
        console.log(`Lookup failed: '${name}'`);
      }
      here++;
    } else if (directive === "d") {
      here++;
    } else if (directive === "s") {
      here += Buffer.byteLength(line.slice(2), "utf8") + 1;
    } else if (directive === "D") {
      const start = here;
      const [name, label, className] = parseDictionaryLine(line);
      const xt = lookup(label);
      const classHandler = lookup(className);
      if (xt === -1) {
        console.log(`Lookup failed: '${label}'`);
      }
      if (classHandler === -1) {
        console.log(`Lookup failed: '${className}'`);
      }
      image.set([
        dictionaryEntries === 0 ? 0 : lastDictionaryEntry,
        xt, classHandler, 0, 0, 0, 0, 0, 0,
      ], start);
      lastDictionaryEntry = start;
      dictionaryEntries++;
      here += dictionaryEntrySize(name);
    }
  });
}

function save(filename) {
  const output = Buffer.alloc(here * 4);
  for (let index = 0; index < here; index++) {
    output.writeInt32LE(image[index], index * 4);
  }
  fs.writeFileSync(filename, output);
}

function main() {
  if (process.argv.length < 3) {
    console.log("No file specified.");
    process.exitCode = 1;
    return;
  }
  const filename = process.argv[2];
  pass1(filename);
  pass2(filename);
  pass3(filename);
  pass4(filename);
  pass5(filename);
  save("ngaImage");
  console.log(`Wrote ${here} cells to ngaImage`);
}

try {
  main();
} catch (error) {
  console.log(error.message);
  process.exitCode = 1;
}
